src/validation.cpp

```diff
@@ -4902,36 +4916,26 @@ void ChainstateManager::CheckBlockIndex()
         if (!pindex->HaveNumChainTxs()) assert(pindex->nSequenceId <= 0); // nSequenceId can't be set positive for blocks that aren't linked (negative is used for preciousblock)
         // VALID_TRANSACTIONS is equivalent to nTx > 0 for all nodes (whether or not pruning has occurred).
         // HAVE_DATA is only equivalent to nTx > 0 (or VALID_TRANSACTIONS) if no pruning has occurred.
-        // Unless these indexes are assumed valid and pending block download on a
-        // background chainstate.
-        if (!m_blockman.m_have_pruned && !pindex->IsAssumedValid()) {
+        if (!m_blockman.m_have_pruned) {
             // If we've never pruned, then HAVE_DATA should be equivalent to nTx > 0
             assert(!(pindex->nStatus & BLOCK_HAVE_DATA) == (pindex->nTx == 0));
-            if (pindexFirstAssumeValid == nullptr) {
-                // If we've got some assume valid blocks, then we might have
-                // missing blocks (not HAVE_DATA) but still treat them as
-                // having been processed (with a fake nTx value). Otherwise, we
-                // can assert that these are the same.
-                assert(pindexFirstMissing == pindexFirstNeverProcessed);
-            }
+            assert(pindexFirstMissing == pindexFirstNeverProcessed);
         } else {
             // If we have pruned, then we can only say that HAVE_DATA implies nTx > 0
             if (pindex->nStatus & BLOCK_HAVE_DATA) assert(pindex->nTx > 0);
         }
         if (pindex->nStatus & BLOCK_HAVE_UNDO) assert(pindex->nStatus & BLOCK_HAVE_DATA);
         if (pindex->IsAssumedValid()) {
-            // Assumed-valid blocks should have some nTx value.
-            assert(pindex->nTx > 0);
             // Assumed-valid blocks should connect to the main chain.
             assert((pindex->nStatus & BLOCK_VALID_MASK) >= BLOCK_VALID_TREE);
-        } else {
-            // Otherwise there should only be an nTx value if we have
-            // actually seen a block's transactions.
-            assert(((pindex->nStatus & BLOCK_VALID_MASK) >= BLOCK_VALID_TRANSACTIONS) == (pindex->nTx > 0)); // This is pruning-independent.
         }
+        // There should only be an nTx value if we have
+        // actually seen a block's transactions.
+        assert(((pindex->nStatus & BLOCK_VALID_MASK) >= BLOCK_VALID_TRANSACTIONS) == (pindex->nTx > 0)); // This is pruning-independent.
         // All parents having had data (at some point) is equivalent to all parents being VALID_TRANSACTIONS, which is equivalent to HaveNumChainTxs().
-        assert((pindexFirstNeverProcessed == nullptr) == pindex->HaveNumChainTxs());
-        assert((pindexFirstNotTransactionsValid == nullptr) == pindex->HaveNumChainTxs());
+        // HaveNumChainTxs will also be set in the assumeutxo snapshot block from snapshot metadata.
+        assert((pindexFirstNeverProcessed == nullptr || pindex == snap_base) == pindex->HaveNumChainTxs());
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505114232

In commit "assumeutxo: Get rid of faked nTx and nChainTx values" (d351ea2f887dc81993799010ac2fb6a220f07561)

> we'll set `nChainTx` (which was previously set to the chainparams value) back to 0

Wow, great catch. It never occurred to me that we would set the `nChainTx` back to 0 after setting it to another value. But that's exactly what we do at the beginning of ReceivedBlockTransactions():

https://github.com/bitcoin/bitcoin/blob/ba907f96ad37c09c49c0e1532fad118fcb8dd4a8/src/validation.cpp#L3616

It looks like this bug was actually present in the previous assert too, and not new in this commit?

In light of that, I guess the check we would need here is:

```c++
if (pindex->HaveNumChainTxs()) {
    assert(pindexFirstNeverProcessed == nullptr || pindex == snap_base);
} else {
    assert(pindexFirstNeverProcessed != nullptr);   
}
```

But I think a better change would be to make ReceivedBlockTransactions() not reset nChainTx to 0 at all. Setting the snapshot block `nChainTx` to a real value, then back to 0, then back to a real value again seems like confusing behavior that could break other code aside from this assert.

Reply:
> It looks like this bug was actually present in the previous assert too, and not new in this commit?

Yes, I encountered this issue testing this branch on signet (and hadn't tried that with master), but it should also fail on master. So I think it would also be ok to fix this not here but in a separate PR.

Slightly related: I wonder if we should do something differently in the situation where the chainparams-provided `nChainTx` turns out to be wrong. Of course that's trusted data and shouldn't be incorrect normally - but if it is anyway I think that we'd currently continue with the inconsistent `nChainTx` data and wouldn't even log an error - unless `-checkblockindex` is enabled, in which case we'd terminate with an assert.

Reply:
> Agree it should probably be a separate PR. 

ok - I can work on a follow-up that fixes the bug (preferably in `ReceivedBlockTransactions` instead of just adjusting the checks) , and add a test case that would trigger the assert on master.

Reply:
> ok - I can work on a follow-up that fixes the bug (preferably in `ReceivedBlockTransactions` instead of just adjusting the checks) , and add a test case that would trigger the assert on master.

That would be great, and I'd be happy to review. Let me know if you think I should update the CheckBlockIndex assert in this PR as described above https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505146598 too here.

I did experiment a little with adding asserts to ReceivedBlockTransactions:

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3589,6 +3589,8 @@ void Chainstate::TryAddBlockIndexCandidate(CBlockIndex* pindex)
 void ChainstateManager::ReceivedBlockTransactions(const CBlock& block, CBlockIndex* pindexNew, const FlatFilePos& pos)
 {
     AssertLockHeld(cs_main);
+    assert(pindexNew->nTx == 0);
+    assert(pindexNew->nChainTx == 0);
     pindexNew->nTx = block.vtx.size();
     pindexNew->nChainTx = 0;
     pindexNew->nFile = pos.nFile;
```

This makes the two assumeutxo tests fail as expected, but it also makes the `rpc_getblockfrompeer.py` test fail the `nTx == 0` check for some reason. Hopefully a straightforward fix ReceivedBlockTransactions should be possible, though.

Reply:
> Let me know if you think I should update the CheckBlockIndex assert in this PR as described above https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505146598 too here.

Up to you, but I think it'd be good to have this fixed and would be happy to re-ACK.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1508080078

> Up to you, but I think it'd be good to have this fixed and would be happy to re-ACK.

Ok, I updated the assertion in a new commit 438ac1ffc23ec5134f2ac74cee75ff145078df09, which could also be a standalone PR. (I'm assuming by "fixed" you mean changing the assertion, not changing ReceivedBlockTransactions, which I think would be harder to do correctly.)

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1508212488

> Ok, I updated the assertion in a new commit [438ac1f](https://github.com/bitcoin/bitcoin/commit/438ac1ffc23ec5134f2ac74cee75ff145078df09), which could also be a standalone PR. (I'm assuming by "fixed" you mean changing the assertion, not changing ReceivedBlockTransactions, which I think would be harder to do correctly.)

New asserts had some problems reported https://github.com/bitcoin/bitcoin/pull/29370#pullrequestreview-1912039200 with a test case triggering failures posted https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. It was possible to fix the asserts in CheckBlockIndex to cope with ReceivedBlockTransactions setting 0 nChainTx values, but the checks started getting more complicated and it turned out to be simpler to just update ReceivedBlockTransactions not to mess up nChainTx instead.

Also debugging the test case revealed that the new crash it was triggering with this PR was not the same as the crash it was triggering before this PR. So commit 438ac1ffc23ec5134f2ac74cee75ff145078df09 would not actually fix anything. The test case works by submitting a snapshot block after loading the snapshot, out of order before previous blocks were downloaded. Before this PR, this would cause ReceivedBlockTransactions to set a fake nChainTx value on the snapshot block and the `pindex->nChainTx == pindex->nTx + prev_chain_tx` check on the next block to fail. After this PR it would cause ReceivedBlockTransactions to set a 0 nChainTx value and the `assert((pindexFirstNeverProcessed == nullptr || pindex == snap_base) == pindex->HaveNumChainTxs());` check to fail.